### ISSUE TITLE
Composition of directory bind controls behaves oddly

### ISSUE THREAD

**Issue by tmager** [apptainer/singularity#2275]
## Version of Singularity:

Current master (`3.0.0-218.g093d5f68`) and 3.0.0 release.

### Expected behavior
When attempting to `--bind` a directory , it would be expected that the directory is bound in the container regardless of options (e.g. `-c`/`--no-home`) that alter what directories are bound by default:
```
~ $ singularity exec /home/tom/ubuntu.sif ls -l /home/tom
total 27356
drwxrwxr-x 2 tom tom     4096 Oct 23 08:43 guest
drwxrwxr-x 2 tom tom     4096 Oct 23 08:43 host
-rwxr-xr-x 1 tom tom 27979776 Oct 23 08:45 ubuntu.sif
< ... etc ... >
~ $ singularity exec -c /home/tom/ubuntu.sif ls -l /home/tom
total 0
~ $ singularity exec -c -B /home/tom/guest /home/tom/ubuntu.sif ls -l /home/tom
total 4
drwxrwxr-x 2 tom tom 4096 Oct 23 12:43 guest
~ $ singularity exec -c -B /home/tom/guest --no-home /home/tom/ubuntu.sif ls -l /home/tom
total 4
drwxrwxr-x 2 tom tom 4096 Oct 23 12:43 guest
```
If my memory serves, this was how these options worked in Singularity 2.x, and I did not see any indication in the documentation that there was an intentional change to this behavior.

### Actual behavior
When the `-c`/`-C` options are used, trying to `--bind` a directory contained inside your home directory only works when specified in conjunction with `--no-home`:
```
~ $ singularity exec /home/tom/ubuntu.sif ls -l /home/tom
total 27356
drwxrwxr-x 2 tom tom     4096 Oct 23 08:43 guest
drwxrwxr-x 2 tom tom     4096 Oct 23 08:43 host
-rwxr-xr-x 1 tom tom 27979776 Oct 23 08:45 ubuntu.sif
< ... etc ... >
~ $ singularity exec -c /home/tom/ubuntu.sif ls -l /home/tom
total 0
~ $ singularity exec -c -B /home/tom/guest /home/tom/ubuntu.sif ls -l /home/tom
total 0
~ $ singularity exec -c -B /home/tom/guest --no-home /home/tom/ubuntu.sif ls -l /home/tom
total 4
drwxrwxr-x 2 tom tom 4096 Oct 23 12:43 guest
```
But if the directory is outside your home directory, it works as expected:
```
~ $ singularity exec -c -B /home/tmp /home/tom/ubuntu.sif ls -l /home/
total 4
drwxr-xr-x 2 root root 4096 Oct 23 12:59 tmp
drwxr-xr-x 2 tom  tom    40 Oct 23 13:01 tom
```

### Steps to reproduce behavior

Build a Singularity image, and try mounting directories as above.

**Comment by jmstover** [comment]
When attempting this (`singularity shell -c -B /home/jason/tmp ubuntu-latest/`), you get:
```
DEBUG   [U=1000,P=32702]   mountGeneric()                
Skipping mount /usr/local/singularity/3.0.0-dist/var/singularity/mnt/session/final/home/jason/tmp,
/usr/local/singularity/3.0.0-dist/var/singularity/mnt/session/final/home/jason already mounted
```

So, it appears that `$HOME` is getting mounted tmpfs from the `-c` option, then a directory from within `$HOME` is attempted to be mounted... But `mountGeneric()` sees `$HOME` as already mounted.

@cclerget @bauerm97 ... would this be a logic change from 2.x behavior?

**Comment by cclerget** [comment]
@tmager @jmstover Yes this is a logic change. Home directory using a staging directory that's why you actually need to specify `--no-home` otherwise the staging directory is mounted and Singularity consider home directory as already mounted when you try to bind something in it.

**Comment by cclerget** [comment]
After testing, the issue seems fixed in 3.4.0. Feel free to re-open if this is not the case
